### PR TITLE
[bugfix] validate AR sampled-token logprob alignment

### PR DESCRIPTION
**Pull request by hbhflw2000**
## Summary

This PR hardens the AR scheduler logprob contract for RL callers.

When a request asks for logprobs, the scheduler now slices the request-local output and verifies that each sampled-token logprob row corresponds exactly to the generated token at the same position.

## Motivation

RL training consumes rollout sampled-token logprobs as old policy logprobs. Returning shifted, incomplete, non-finite, or token-mismatched rows silently corrupts PPO/GRPO statistics and makes downstream parity diagnostics misleading.

The scheduler previously forwarded `logprobs.slice_request(...)` without validating this contract.

## Changes

Add `_slice_sampled_logprobs(...)` in `OmniARScheduler` and use it for AR requests that generated tokens and requested logprobs.

The helper validates:

- model-runner logprobs are present;
- token-id and logprob arrays are rank-2;
- row counts match generated token count;
- sampled-token columns are non-empty;
- the first returned token id in every row equals the generated token id;
- sampled-token logprob values are finite.

Invalid output now fails fast with a descriptive `RuntimeError`.

## Tests

Added unit coverage for:

- request-local slicing with cumulative generated-token offsets;
- preservation of sampled-token values;
- missing model-runner logprobs;
- generated-token/logprob row-count mismatch;
- sampled-token ID mismatch;
- non-finite sampled-token logprobs.

Also validated the AR path locally with TP1 and TP4. The returned sampled-token rows matched generated tokens with no missing, duplicated, or non-finite values.

Repository pre-commit checks passed for the changed files, including ruff, formatting, typos, and pickle-import checks.

## Compatibility

This does not change sampling, logits, or behavior for requests that do not ask for logprobs.

For requests that explicitly request logprobs, invalid runner output now raises immediately instead of propagating corrupted RL statistics.

### CODE REVIEW

**Review by SamitHuang — CHANGES_REQUESTED** [review]
This hardens the AR logprob contract for RL, but it is closer to fail-fast validation than a root-cause bugfix

File: vllm_omni/core/sched/omni_ar_scheduler.py [hunk 1]
```diff
@@ -35,6 +36,47 @@
 logger = init_logger(__name__)
 
 
+def _slice_sampled_logprobs(logprobs: Any, req_index: int, sampled_token_ids: list[int]) -> Any:
+    """Slice and validate the sampled-token logprobs for one AR request."""
+    if logprobs is None:
+        raise RuntimeError("AR logprobs were requested, but the model runner returned none")
```
Comment (SamitHuang) [inline vllm_omni/core/sched/omni_ar_scheduler.py]: Raising a `RuntimeError` here will crash the entire scheduler loop and kill all concurrent requests on the engine. Should we instead abort only the affected request by setting its status to `FINISHED_ABORTED` with an error message, preserving overall engine stability?

File: vllm_omni/core/sched/omni_ar_scheduler.py [hunk 2]
```diff
@@ -35,6 +36,47 @@
 logger = init_logger(__name__)
 
 
+def _slice_sampled_logprobs(logprobs: Any, req_index: int, sampled_token_ids: list[int]) -> Any:
```
Comment (SamitHuang) [inline vllm_omni/core/sched/omni_ar_scheduler.py:43]: This validation detects the symptom (invalid logprobs) and fails fast, but it doesn't fix the root cause in the model runner that is generating the shifted or non-finite logprobs. Is the root cause being addressed in a separate PR, or should the model runner be fixed here to actually unblock PPO/GRPO?

File: tests/core/sched/test_omni_ar_scheduler_logprobs.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        cumulative_rows=[0, 2],
+    )
+
+    sliced = _slice_sampled_logprobs(all_logprobs, req_index=1, sampled_token_ids=[11, 12])
+
+    np.testing.assert_array_equal(sliced.logprob_token_ids[:, 0], [11, 12])
+    np.testing.assert_allclose(sliced.logprobs[:, 0], [-0.11, -0.12])
+
+
+def test_rejects_missing_logprobs() -> None:
```
Comment (SamitHuang) [inline tests/core/sched/test_omni_ar_scheduler_logprobs.py:46]: Missing regression test for this bug fix at the scheduler level. While the helper is unit-tested, please add a test that reproduces the original issue by mocking `OmniARScheduler.update_from_output` with bad logprobs and verifies the scheduler correctly handles the invalid runner output.

**Review by SamitHuang — CHANGES_REQUESTED** [review]
The PR introduces validation to prevent silent corruption of RL statistics. However, there are architectural and testing gaps that need to be addressed, specifically regarding error handling impact on the engine and missing scheduler-level regression tests.

File: vllm_omni/core/sched/omni_ar_scheduler.py (hunk 1 above)
Comment (hbhflw2000) [reply vllm_omni/core/sched/omni_ar_scheduler.py]: Implemented in bdb511d8. Validation now runs before request token state is mutated. A contract violation is isolated to that request: it is marked FINISHED_ERROR, gets the validation message in stop_reason, and emits an error finish reason. The scheduler continues processing remaining requests in the same batch. I used FINISHED_ERROR rather than FINISHED_ABORTED because this is an engine-side invalid runner response, not a client cancellation.

File: vllm_omni/core/sched/omni_ar_scheduler.py (hunk 2 above)
Comment (hbhflw2000) [reply vllm_omni/core/sched/omni_ar_scheduler.py:43]: Agreed. This PR is intentionally scoped as fail-fast validation plus request-local containment, not the model-runner root-cause fix. It makes the corrupted sampled-token logprob contract observable without taking down unrelated requests. The underlying multi-node runner/logprob parity investigation remains separate; this change is safe independently because it protects PPO/GRPO from silently consuming malformed data.

File: tests/core/sched/test_omni_ar_scheduler_logprobs.py (hunk 3 above)
Comment (hbhflw2000) [reply tests/core/sched/test_omni_ar_scheduler_logprobs.py:46]: Added in bdb511d8. test_invalid_logprobs_finish_only_the_affected_scheduler_request mocks OmniARScheduler.update_from_output with one malformed and one valid request in the same model-runner output. It verifies the bad request is finished with FinishReason.ERROR, is not token-mutated, is freed, and the valid request still receives its token and sampled logprob.

**Review by SamitHuang — APPROVED** [review]
The added SampledLogprobContractError helper LGTM
